### PR TITLE
JIT: Generalize check for cycles when finding loops

### LINKED ISSUE #96145: [runtime-jit-experimental] Assertion failed 'loops->NumLoops() == 0'
`runtime-jit-experimental` is blocked by the following assert:

```
Assert failure(PID 976 [0x000003d0], Thread: 3776 [0x0ec0]): Assertion failed 'loops->NumLoops() == 0' in 'hello_leaveintotrybody_leaves_il.Class1:Main():int' during 'Place 'align' instructions' (IL size 244; hash 0xc0639335; Tier1-OSR)

    File: D:\a\_work\1\s\src\coreclr\jit\compiler.cpp Line: 5436
    Image: C:\h\w\B03D09B3\p\corerun.exe
```

[Example run](https://dev.azure.com/dnceng-public/public/_build/results?buildId=502570&view=results)

cc @dotnet/jit-contrib

### PR DESCRIPTION
**Pull request by jakobbotsch**
`optFindNewLoops` tries to leave a breadcrumb for loop alignment about whether it needs to bother with identifying loops. It does so by checking for whether any natural loops were found, but also whether any irreducible loops were found (i.e. any cycle). The idea is that optimizations that run between finding loops and loop alignment can easily remove edges that make irreducible loops into natural loops.

However, the check for "any cycle" was not precise enough; it made the assumption that `NumLoops() > 0 || NumImproperLoopHeaders > 0` was equivalent to there being a cycle in the flow graph. But that is not true, specifically because we only consider blocks to be loop headers when their backedges are non-exceptional. Odd loop structures can exist that iterate through required exception throws, like in the case of #96145. In that case we have a cycle through the handler, which ends up being a natural loop after some optimizations run.  ([Flowgraph](https://gist.github.com/jakobbotsch/61d074c95127b1c9ba46b8ff10ef2a36) -- the cycle is `BB08 -> BB09 -> BB12 -> BB05 -> BB08`. The entry edge would be `BB02 -> BB08`.). But since the backedge is exceptional, we do not consider it as a loop during loop finding, and we do not consider it as an improper loop header either.

To fix the situation we generalize `fgComputeDfs` to identify when the flow graph has a cycle, which is the fully general property we want for the breadcrumb. As a bonus we can skip identifying loops if we know the flow graph has no cycles.

Fix #96145

### CODE REVIEW

File: src/coreclr/jit/flowgraph.cpp [hunk 1]
```diff
@@ -4388,13 +4403,20 @@ FlowGraphNaturalLoops* FlowGraphNaturalLoops::Find(const FlowGraphDfsTree* dfsTr
     {
         unsigned          rpoNum = dfsTree->GetPostOrderCount() - i;
         BasicBlock* const block  = dfsTree->GetPostOrder(i - 1);
-        JITDUMP("%02u -> " FMT_BB "[%u, %u]\n", rpoNum + 1, block->bbNum, block->bbPreorderNum + 1,
-                block->bbNewPostorderNum + 1);
+        JITDUMP("%02u -> " FMT_BB "[%u, %u]\n", rpoNum, block->bbNum, block->bbPreorderNum, block->bbNewPostorderNum);
```
Comment (jakobbotsch) [inline src/coreclr/jit/flowgraph.cpp:4406]: I made this output 1-based to exactly match the original loop finding way back when I originally factored it out and was investigating a cause of diffs, but forgot to revert it back to be 0-based.

**Comment by jakobbotsch** [comment]
Originally I looked at simply generalizing loop finding to actually find the loop in the example above, which amounts to switching to use `BlockPredsWithEH` here:
https://github.com/dotnet/runtime/blob/bdd9a4736b1648b90d771e7c2fb1f0d3d3a0f754/src/coreclr/jit/flowgraph.cpp#L4408

This also fixes the problem, but the loop structure is so odd that I don't think we want to consider it a loop. One odd thing, for instance, is that it is impossible to canonicalize the loop to have a preheader.

**Comment by jakobbotsch** [comment]
cc @dotnet/jit-contrib PTAL @AndyAyersMS @BruceForstall

**Review by AndyAyersMS — APPROVED** [review]

**Review by BruceForstall — APPROVED** [review]

**Pull request merged**